proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,146 @@
+
+1. *Infer* the type of `e_0`, which we shall call `sigma_0`,
+2. Expose the first *n* argument types by instantiating `sigma_0` into a type of the form `sigma_1 -> ... -> sigma_n -> sigma_result`,
+3. *Check* each argument expression *e_i* against the corresponding type `sigma_i`,
+4. If we are in checking mode, run the subtype check (`tcSubType`) of `sigma_result` against the type being pushed.
+
+For example, suppose we are inferring the type of `id True`:
+
+1. We infer the type of `id` by merely looking up in the environment, `forall a. a -> a`,
+2. We expose one argument, and for that we instantiate the type above with a new type variable, leading to `alpha -> alpha`,
+3. Now we check the argument `True` against `alpha`:
+   a. We infer the type of `True`, which is `Bool`,
+   b. Steps (2) and (3) are not needed here, since there are no arguments,
+   c. We perform the subtype check `alpha <= Bool`, leading to the unification `alpha := Bool`;
+4. The result of the inference is `alpha`; after zonking the result is `Bool`.
+
+When `ImpredicativeTypes` is on we introduce an additional step between (2) and (3), which we call *quick look*. Quick look traverses the arguments of the application, and tries to infer as much impredicative instantiations as possible. The results are applied before step (3), which means that arguments are type checked with the correct impredicative instantiations. The "quick look" pass only traverses *simple expressions*, which we define to be (possibly nested) applications of variables in the environment.
+
+Another way to look at this proposal is that each application is type checked *twice*: first the "quick look" pass tries to infer impredicative instantiation, which is then fed to the second, real pass. For simple expressions it is crystal clear how impredicative instantiation is threaded. On the contrary, it never looks at abstractions, pattern matching, `let`s, or any other expression.
+
+One important feature of Haskell's type system that "quick look" uses is the invariance of type constructors. In short, the subsumption rules ensure that if, for example, `Maybe t` is a more polymorphic than `Maybe s`, it must be the case that `t` equals `s`. This property holds for every type constructor except for function types, which require a slightly more complex handling. The proposed inference algorithm never tries to perform any complicated analysis on other types: impredicativity must be the *only obvious* solution to make the program type check (or as we say in the paper, "impredicativity is never guessed"). Note that since [proposal 287, *Simplify Subsumption*](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0287-simplify-subsumption.rst) has been accepted, function types are also considered invariant.
```

Comment:
Is it actually the case that function types require more complex handling? Or are they handled uniformly thanks to #287 ?

Reply:
I have removed the sentence about arrows. As you say, since #287 they are treated as any other type.